File: app/src/main/java/com/veganbeauty/admin/features/product/ProductViewModel.java
package com.veganbeauty.admin.features.product;

import android.app.Application;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.MutableLiveData;

import com.veganbeauty.admin.data.local.RootieAdminDatabase;
import com.veganbeauty.admin.data.local.entities.EntityUtils;
import com.veganbeauty.admin.data.local.entities.ProductEntity;
import com.veganbeauty.admin.data.remote.FirebaseService;
import com.veganbeauty.admin.data.repository.ProductRepository;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class ProductViewModel extends AndroidViewModel {

    private final ProductRepository repository;
    private final LiveData<List<ProductEntity>> allProducts;

    // LiveData for search query, active category, and sort method
    private final MutableLiveData<String> searchQuery = new MutableLiveData<>("");
    private final MutableLiveData<String> selectedCategory = new MutableLiveData<>("");
    private final MutableLiveData<Set<String>> selectedSubcategories = new MutableLiveData<>(Collections.emptySet());
    private final MutableLiveData<String> filterStockStatus = new MutableLiveData<>("ALL"); // ALL, IN_STOCK, OUT_OF_STOCK
    private final MutableLiveData<String> filterHiddenStatus = new MutableLiveData<>("ALL"); // ALL, HIDDEN, VISIBLE
    private final MutableLiveData<String> sortOrder = new MutableLiveData<>("DEFAULT"); // DEFAULT, PRICE_ASC, PRICE_DESC, NAME_ASC, NAME_DESC

    private final MediatorLiveData<List<ProductEntity>> filteredProducts = new MediatorLiveData<>();

    public ProductViewModel(@NonNull Application application) {
        super(application);
        RootieAdminDatabase database = RootieAdminDatabase.getDatabase(application);
        repository = new ProductRepository(database.productDao(), new FirebaseService());
        allProducts = repository.getAllProducts();

        // Combine sources for filtered products
        filteredProducts.addSource(allProducts, products -> updateFilteredProducts());
        filteredProducts.addSource(searchQuery, query -> updateFilteredProducts());
        filteredProducts.addSource(selectedCategory, category -> updateFilteredProducts());
        filteredProducts.addSource(selectedSubcategories, subs -> updateFilteredProducts());
        filteredProducts.addSource(filterStockStatus, stock -> updateFilteredProducts());
        filteredProducts.addSource(filterHiddenStatus, hidden -> updateFilteredProducts());
        filteredProducts.addSource(sortOrder, sort -> updateFilteredProducts());
    }

    public LiveData<List<ProductEntity>> getAllProducts() {
        return allProducts;
    }

    public MutableLiveData<String> getSearchQuery() {
        return searchQuery;
    }

    public MutableLiveData<String> getSelectedCategory() {
        return selectedCategory;
    }

    public MutableLiveData<Set<String>> getSelectedSubcategories() {
        return selectedSubcategories;
    }

    public MutableLiveData<String> getFilterStockStatus() {
        return filterStockStatus;
    }

    public MutableLiveData<String> getFilterHiddenStatus() {
        return filterHiddenStatus;
    }

    public MutableLiveData<String> getSortOrder() {
        return sortOrder;
    }

    public LiveData<List<ProductEntity>> getFilteredProducts() {
        return filteredProducts;
    }

    public void syncFromFirebase() {
        new Thread(() -> {
            try {
                repository.checkAndSeedProducts(getApplication());
            } catch (Exception e) {
                e.printStackTrace();
            }
        }).start();
    }

    public void deleteProduct(ProductEntity product) {
        new Thread(() -> {
            try {
                repository.deleteProduct(product);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }).start();
    }

    public void saveProduct(ProductEntity product) {
        new Thread(() -> {
            try {
                repository.saveProduct(product);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }).start();
    }

    public void toggleProductVisibility(ProductEntity product) {
        ProductEntity updatedProduct = EntityUtils.copy(product, !product.isHidden());
        saveProduct(updatedProduct);
    }

    private void updateFilteredProducts() {
        List<ProductEntity> products = allProducts.getValue();
        if (products == null) {
            products = new ArrayList<>();
        }
        String query = searchQuery.getValue();
        query = query != null ? query.trim().toLowerCase() : "";
        String category = selectedCategory.getValue();
        if (category == null) category = "";
        Set<String> subcategories = selectedSubcategories.getValue();
        if (subcategories == null) subcategories = new HashSet<>();
        String stockStatus = filterStockStatus.getValue();
        if (stockStatus == null) stockStatus = "ALL";
        String hiddenStatus = filterHiddenStatus.getValue();
        if (hiddenStatus == null) hiddenStatus = "ALL";
        String sort = sortOrder.getValue();
        if (sort == null) sort = "DEFAULT";

        List<ProductEntity> result = new ArrayList<>();

        for (ProductEntity product : products) {
            // 1. Search Filter
            if (!query.isEmpty()) {
                String name = product.getName() != null ? product.getName().toLowerCase() : "";
                String sku = product.getSku() != null ? product.getSku().toLowerCase() : "";
                if (!name.contains(query) && !sku.contains(query)) {
                    continue;
                }
            }

            // 2. Category Filter
            if (!category.isEmpty()) {
                String cat = product.getCategory();
                if (cat == null || !cat.equalsIgnoreCase(category)) {
                    continue;
                }
            }

            // 3. Subcategories Filter
            if (!subcategories.isEmpty()) {
                String sub = product.getSubcategory();
                if (sub == null) {
                    continue;
                }
                String[] parts = sub.split(",");
                boolean subMatch = false;
                for (String part : parts) {
                    if (subcategories.contains(part.trim())) {
                        subMatch = true;
                        break;
                    }
                }
                if (!subMatch) {
                    continue;
                }
            }

            // 4. Stock Status Filter
            if ("IN_STOCK".equals(stockStatus)) {
                if (product.getStock() <= 0) {
                    continue;
                }
            } else if ("OUT_OF_STOCK".equals(stockStatus)) {
                if (product.getStock() != 0) {
                    continue;
                }
            }

            // 5. Hidden Status Filter
            if ("HIDDEN".equals(hiddenStatus)) {
                if (!product.isHidden()) {
                    continue;
                }
            } else if ("VISIBLE".equals(hiddenStatus)) {
                if (product.isHidden()) {
                    continue;
                }
            }

            result.add(product);
        }

        // 6. Sorting
        final String finalSort = sort;
        result.sort((p1, p2) -> {
            switch (finalSort) {
                case "PRICE_ASC":
                    return Long.compare(p1.getPrice(), p2.getPrice());
                case "PRICE_DESC":
                    return Long.compare(p2.getPrice(), p1.getPrice());
                case "NAME_ASC":
                    String n1 = p1.getName() != null ? p1.getName() : "";
                    String n2 = p2.getName() != null ? p2.getName() : "";
                    return n1.compareTo(n2);
                case "NAME_DESC":
                    String nd1 = p1.getName() != null ? p1.getName() : "";
                    String nd2 = p2.getName() != null ? p2.getName() : "";
                    return nd2.compareTo(nd1);
                default:
                    return 0;
            }
        });

        filteredProducts.setValue(result);
    }
}

File: app/src/main/java/com/veganbeauty/admin/features/product/ProductViewModel.kt
package com.veganbeauty.admin.features.product

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.veganbeauty.admin.data.local.RootieAdminDatabase
import com.veganbeauty.admin.data.local.entities.ProductEntity
import com.veganbeauty.admin.data.remote.FirebaseService
import com.veganbeauty.admin.data.repository.ProductRepository
import kotlinx.coroutines.launch

class ProductViewModel(application: Application) : AndroidViewModel(application) {

    private val repository: ProductRepository
    val allProducts: LiveData<List<ProductEntity>>

    // LiveData for search query, active category, and sort method
    val searchQuery = MutableLiveData<String>("")
    val selectedCategory = MutableLiveData<String>("")
    val selectedSubcategories = MutableLiveData<Set<String>>(emptySet())
    val filterStockStatus = MutableLiveData<String>("ALL") // ALL, IN_STOCK, OUT_OF_STOCK
    val filterHiddenStatus = MutableLiveData<String>("ALL") // ALL, HIDDEN, VISIBLE
    val sortOrder = MutableLiveData<String>("DEFAULT") // DEFAULT, PRICE_ASC, PRICE_DESC, NAME_ASC, NAME_DESC

    private val _filteredProducts = MediatorLiveData<List<ProductEntity>>()
    val filteredProducts: LiveData<List<ProductEntity>> get() = _filteredProducts

    init {
        val database = RootieAdminDatabase.getDatabase(application)
        repository = ProductRepository(database.productDao(), FirebaseService())
        allProducts = repository.allProducts

        // Combine sources for filtered products
        _filteredProducts.addSource(allProducts) { updateFilteredProducts() }
        _filteredProducts.addSource(searchQuery) { updateFilteredProducts() }
        _filteredProducts.addSource(selectedCategory) { updateFilteredProducts() }
        _filteredProducts.addSource(selectedSubcategories) { updateFilteredProducts() }
        _filteredProducts.addSource(filterStockStatus) { updateFilteredProducts() }
        _filteredProducts.addSource(filterHiddenStatus) { updateFilteredProducts() }
        _filteredProducts.addSource(sortOrder) { updateFilteredProducts() }
    }

    fun syncFromFirebase() {
        viewModelScope.launch {
            try {
                repository.checkAndSeedProducts(getApplication())
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun deleteProduct(product: ProductEntity) {
        viewModelScope.launch {
            try {
                repository.deleteProduct(product)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun saveProduct(product: ProductEntity) {
        viewModelScope.launch {
            try {
                repository.saveProduct(product)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun toggleProductVisibility(product: ProductEntity) {
        val updatedProduct = product.copy(isHidden = !product.isHidden)
        saveProduct(updatedProduct)
    }

    private fun updateFilteredProducts() {
        val products = allProducts.value ?: emptyList()
        val query = searchQuery.value?.trim()?.lowercase() ?: ""
        val category = selectedCategory.value ?: ""
        val subcategories = selectedSubcategories.value ?: emptySet()
        val stockStatus = filterStockStatus.value ?: "ALL"
        val hiddenStatus = filterHiddenStatus.value ?: "ALL"
        val sort = sortOrder.value ?: "DEFAULT"

        var result = products

        // 1. Search Filter
        if (query.isNotEmpty()) {
            result = result.filter {
                it.name.lowercase().contains(query) || it.sku.lowercase().contains(query)
            }
        }

        // 2. Category Filter
        if (category.isNotEmpty()) {
            result = result.filter {
                it.category.equals(category, ignoreCase = true)
            }
        }

        // 3. Subcategories Filter
        if (subcategories.isNotEmpty()) {
            result = result.filter { product ->
                val productSubs = product.subcategory.split(",").map { it.trim() }.filter { it.isNotEmpty() }
                productSubs.any { it in subcategories }
            }
        }

        // 4. Stock Status Filter
        result = when (stockStatus) {
            "IN_STOCK" -> result.filter { it.stock > 0 }
            "OUT_OF_STOCK" -> result.filter { it.stock == 0 }
            else -> result
        }

        // 5. Hidden Status Filter
        result = when (hiddenStatus) {
            "HIDDEN" -> result.filter { it.isHidden }
            "VISIBLE" -> result.filter { !it.isHidden }
            else -> result
        }

        // 6. Sorting
        result = when (sort) {
            "PRICE_ASC" -> result.sortedBy { it.price }
            "PRICE_DESC" -> result.sortedByDescending { it.price }
            "NAME_ASC" -> result.sortedBy { it.name }
            "NAME_DESC" -> result.sortedByDescending { it.name }
            else -> result // DEFAULT
        }

        _filteredProducts.value = result
    }
}
